File: recyclerview-adapter-binding/src/main/java/cn/ads/recyclerview/AdsBindingAdapter.java
package cn.ads.recyclerview;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import java.util.List;

/**
 * @author QiuJie
 */
public abstract class AdsBindingAdapter<T, VB extends ViewBinding> extends RecyclerView.Adapter<AdsBindingViewHolder<VB>> {
    @Nullable
    private List<T> mList;

    public AdsBindingAdapter() {
    }

    public AdsBindingAdapter(@Nullable List<T> list) {
        this.mList = list;
    }

    @Override
    public void onBindViewHolder(@NonNull AdsBindingViewHolder<VB> holder, int position) {
        T item = getItem(position);
        if (item != null) {
            onBindViewHolder(holder, position, item);
        }
    }

    public abstract void onBindViewHolder(@NonNull AdsBindingViewHolder<VB> holder, int position, T item);

    public void submitList(@Nullable final List<T> newList) {
        this.mList = newList;
    }

    public List<T> getCurrentList() {
        return mList;
    }

    @Override
    public int getItemCount() {
        return mList == null ? 0 : mList.size();
    }

    protected T getItem(int position) {
        if (mList != null) {
            return mList.get(position);
        }
        return null;
    }
}

File: ads/src/main/kotlin/ads/view/recyclerview/AdsBindingAdapter.kt
package ads.view.recyclerview

import androidx.recyclerview.widget.AsyncDifferConfig
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.viewbinding.ViewBinding

/**
 * @author QiuJie
 */
abstract class AdsBindingAdapter<T, VB : ViewBinding> : ListAdapter<T, AdsBindingViewHolder<VB>> {
    protected constructor(diffCallback: DiffUtil.ItemCallback<T>) : super(diffCallback)
    protected constructor(config: AsyncDifferConfig<T>) : super(config)

    override fun onBindViewHolder(holder: AdsBindingViewHolder<VB>, position: Int) {
        getItem(position)?.apply { onBindViewHolder(holder, holder.binding, this) }
    }

    abstract fun onBindViewHolder(holder: AdsBindingViewHolder<VB>, binding: VB, item: T)
}